File: LeetCode/Unknown/백준/Gold/16724. 피리 부는 사나이/피리 부는 사나이.cc
#include <iostream>
#include <string>
#include <vector>

using namespace std;

const int NOT = 0;  // 아직 방문하지 않은 곳
const int ING = 1;  // 현재 방문 중인 곳
const int DONE = 2; // 이전에 이미 방문한 곳

int N, M;
vector<int> graph;
vector<int> state;

bool dfs(int cur) {
    if (state[cur] == ING) return true;     // 현재 경로에서 다시 등장 -> 사이클
    if (state[cur] == DONE) return false;   // 이미 탐색 완료
    state[cur] = ING;
    
    bool ret = dfs(graph[cur]);
    state[cur] = DONE;
    return ret;
}

int main() {
    cin >> N >> M;
    vector<string> board(N);
    for (int i = 0; i < N; i++) cin >> board[i];

    graph = vector<int>(N * M);
    state = vector<int>(N * M, 0);

    for (int y = 0; y < N; y++) {
        for (int x = 0; x < M; x++) {
            int idx = y * M + x;
            if (board[y][x] == 'D') graph[idx] = (y + 1) * M + x;
            if (board[y][x] == 'U') graph[idx] = (y - 1) * M + x;
            if (board[y][x] == 'L') graph[idx] = y * M + (x - 1);
            if (board[y][x] == 'R') graph[idx] = y * M + (x + 1);
        }
    }

    int answer = 0;
    for (int i = 0; i < N * M; i++) {
        if (state[i] == 0 && dfs(i)) answer++;
    }

    cout << answer;

    return 0;
}

File: LeetCode/Unknown/백준/Gold/16724. 피리 부는 사나이/피리 부는 사나이.java

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

    static int n, m;
    static char[][] board;

    public static void main(String[] arg) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());

        board = new char[n][m];
        for (int y = 0; y < n; y++) {
            String line = br.readLine();
            for (int x = 0; x < m; x++) {
                board[y][x] = line.charAt(x);
            }
        }

        /*
            there are three state.
            0 = Not visited
            1 = Ing
            2 = Complete
         */

        int[][] state = new int[n][m];
        for (int y = 0; y < n; y++) {
            for (int x = 0; x < m; x++) {
                state[y][x] = 0;
            }
        }

        int answer = 0;
        for (int y = 0; y < n; y++) {
            for (int x = 0; x < m; x++) {
                if (state[y][x] == 0) {
                    if (dfs(state, y, x)) {
                        answer++;
                    }
                }
            }
        }

        System.out.println(answer);
    }

    private static boolean dfs(int[][] state, int y, int x) {
        if (state[y][x] == 2) {
            return false;
        }
        if (state[y][x] == 1) {
            return true;
        }

        int y_ = 0, x_ = 0;
        if (board[y][x] == 'D') {
            y_++;
        }
        if (board[y][x] == 'U') {
            y_--;
        }
        if (board[y][x] == 'R') {
            x_++;
        }
        if (board[y][x] == 'L') {
            x_--;
        }

        state[y][x] = 1;
        boolean flag = dfs(state, y + y_, x + x_);
        state[y][x] = 2;
        return flag;
    }

}
